File: shopping-cart/csharp/src/ShoppingCart/Quantity.cs
namespace ShoppingCart;

public readonly record struct Quantity
{
    public int Value { get; }

    public Quantity(int value)
    {
        if (value <= 0)
        {
            throw new ArgumentOutOfRangeException(nameof(value), "Quantity must be a positive whole number");
        }
        Value = value;
    }

    public static implicit operator int(Quantity q) => q.Value;

    public override string ToString() => Value.ToString();
}

File: shopping-cart/typescript/src/Quantity.ts
export class Quantity {
  readonly value: number;

  constructor(value: number) {
    if (!Number.isInteger(value) || value <= 0) {
      throw new RangeError('Quantity must be a positive whole number');
    }
    this.value = value;
  }

  plus(other: Quantity): Quantity { return new Quantity(this.value + other.value); }
}
